File: 1463.Cherry Pickup II.rb
# [Time Limit Exceeded, due to inefficiency of Ruby Hash]
# @param {Integer[][]} grid
# @return {Integer}
def cherry_pickup(grid)
    def search(i, j1, j2, grid, dp)
        return 0 if i >= grid.size || !(0 <= j1 && j1 < grid[0].size) || !(0 <= j2 && j2 < grid[0].size)
        j1, j2 = j2, j1 if j1 > j2
        if !dp.include? [i, j1, j2]
            if j1 == j2
                pick = grid[i][j1]
            else
                pick = grid[i][j1] + grid[i][j2]
            end
            mx = 0
            for next_j1 in (j1 - 1 .. j1 + 1)
                for next_j2 in (j2 - 1 .. j2 + 1)
                    mx = [mx, search(i + 1, next_j1, next_j2, grid, dp)].max
                end
            end
            dp[[i, j1, j2]] = pick + mx
        end
        dp[[i, j1, j2]]
    end
    search(0, 0, grid[0].size - 1, grid, {})
end

File: 1463.Cherry Pickup II.py
import functools
class Solution:
    def cherryPickup(self, grid: List[List[int]]) -> int:
        @functools.lru_cache(None)
        def dp(i, j1, j2):
            if not (i < len(grid) and 0 <= j1 < len(grid[0]) and 0 <= j2 < len(grid[0])):
                return 0
            if j1 == j2:
                pick = grid[i][j1]
            else:
                pick = grid[i][j1] + grid[i][j2]
            res = 0
            for d1 in (-1, 0, 1):
                for d2 in (-1, 0, 1):
                    res = max(res, dp(i + 1, j1 + d1, j2 + d2))
            return pick + res
        return dp(0, 0, len(grid[0]) - 1)
